src/modules/wlr/workspace_manager.cpp

```diff
@@ -0,0 +1,323 @@
+  for (auto &group : groups_) {
+    group->handle_done();
+  }
+  dp.emit();
+}
+
+auto WorkspaceManager::update() -> void {
+  for (auto &group : groups_) {
+    group->update();
+  }
+  AModule::update();
+}
+
+WorkspaceManager::~WorkspaceManager() {
+  if (!workspace_manager_) {
+    return;
+  }
+
+  zwlr_workspace_manager_v1_destroy(workspace_manager_);
+  workspace_manager_ = nullptr;
+}
+auto WorkspaceManager::remove_workspace_group(uint32_t id) -> void {
+  auto it = std::find_if(groups_.begin(),
+                         groups_.end(),
+                         [id](const std::unique_ptr<WorkspaceGroup> &g) { return g->id() == id; });
+
+  if (it == groups_.end()) {
+    spdlog::warn("Can't find group with id {}", id);
+    return;
+  }
+
+  groups_.erase(it);
+}
+auto WorkspaceManager::commit() -> void { zwlr_workspace_manager_v1_commit(workspace_manager_); }
+
+WorkspaceGroup::WorkspaceGroup(const Bar &bar, Gtk::Box &box, const Json::Value &config,
+                               WorkspaceManager &              manager,
+                               zwlr_workspace_group_handle_v1 *workspace_group_handle, uint32_t id)
+    : bar_(bar),
+      box_(box),
+      config_(config),
+      workspace_manager_(manager),
+      workspace_group_handle_(workspace_group_handle),
+      id_(id) {
+  add_workspace_group_listener(workspace_group_handle, this);
+  auto config_sort_by_name = config_["sort-by-name"];
+  if (config_sort_by_name.isBool()) {
+    sort_by_name = config_sort_by_name.asBool();
+  }
+
+  auto config_sort_by_coordinates = config_["sort-by-coordinates"];
+  if (config_sort_by_coordinates.isBool()) {
+    sort_by_coordinates = config_sort_by_coordinates.asBool();
+  }
+}
+auto WorkspaceGroup::add_button(Gtk::Button &button) -> void {
+  box_.pack_start(button, false, false);
+}
+
+WorkspaceGroup::~WorkspaceGroup() {
+  if (!workspace_group_handle_) {
+    return;
+  }
+
+  zwlr_workspace_group_handle_v1_destroy(workspace_group_handle_);
+  workspace_group_handle_ = nullptr;
+}
+
+auto WorkspaceGroup::handle_workspace_create(zwlr_workspace_handle_v1 *workspace) -> void {
+  auto new_id = ++workspace_global_id;
+  workspaces_.push_back(std::make_unique<Workspace>(bar_, config_, *this, workspace, new_id));
+  spdlog::debug("Workspace {} created", new_id);
+}
+
+auto WorkspaceGroup::handle_remove() -> void {
+  zwlr_workspace_group_handle_v1_destroy(workspace_group_handle_);
+  workspace_group_handle_ = nullptr;
+  workspace_manager_.remove_workspace_group(id_);
+}
+
+auto WorkspaceGroup::handle_output_enter(wl_output *output) -> void {
+  spdlog::debug("Output {} assigned to {} group", (void *)output, id_);
+  for (auto &workspace : workspaces_) {
+    workspace->show();
+  }
+  output_ = output;
+}
+
+auto WorkspaceGroup::handle_output_leave() -> void {
+  spdlog::debug("Output {} remove from {} group", (void *)output_, id_);
+  output_ = nullptr;
+  for (auto &workspace : workspaces_) {
+    workspace->hide();
+  }
+}
```

Comment:
Done.
Check only equality with the current output since compositor assign outputs to group by self. So all_output parameter isn't actual for wlr_workspaces.